packages/compiler/server/serverlib.ts

```diff
@@ -410,8 +410,11 @@ export function createServer(host: ServerHost): Server {
 
   async function getFoldingRanges(params: FoldingRangeParams): Promise<FoldingRange[]> {
     const file = await compilerHost.readFile(getPath(params.textDocument));
-    const ast = parse(file);
+    const ast = parse(file, { comments: true });
     const ranges: FoldingRange[] = [];
+    for (let i = 0; i < ast.comments.length; i++) {
+      addRange(ast.comments[i].pos, ast.comments[i].end);
```

Comment:
Does this come out as three separate comments or one:

// This is a long comment, that does not use /* */
// Yada yada yada
// Etc. etc.

I think we want the same behavior as TS here:

![folding2](https://user-images.githubusercontent.com/75470/178018137-13371b91-b1c1-4419-8c08-5ff8850fc4f2.gif)

I wonder if this is or should be handled when we build ast.comments instead of here. @timotheeguerin thoughts?

Reply:
Looking at the way typescript does it they don't seem to combine them in the ast tree. I think it would be tricky, specially if you have blank line in between(this is something that still get folded)
https://ts-ast-viewer.com/#code/PTAEDMHtIWAKBKARgQwE73ogdgUwB4AuoANgJZ5A

Reply:
Hmm, ok. It would have been easier if the parser combined it. And if it's somebody else's job to collpase these here, then it would also be helpful if we had an parser made avaialable whitespace trivia, not just comments. Maybe we can take this improvement for /* */ and work on that separately?

What do you think about having it be `trivia` instead of comments? true and having another kind for whitespace. Then it wouldn't be too hard to say "if I see a single line comment, end this at the start of the first non-single line comment, non-whitepaces after this." It would be similar to how we collapse decorators together.

Reply:
We could still make helpers like getLeadingCommentRanges/etc. as we need them, but it feels like we want all the trivia ranges as the most primitive output from the parser.

Reply:
The way typescript seems to do it is have an helper method that gets the leadingComments for a node and then iterate of them to combine consecutive single line comments. https://github.com/microsoft/TypeScript/blob/28dc248e5c500c7be9a8c3a7341d303e026b023f/src/services/outliningElementsCollector.ts#L117

We currently count on prettier to do the comment to node assignment

Reply:
Do we need the node mapping for this feature? It seems OK to just collapse all comments that are more than one line separately from collapsing nodes. But we do need to know if there is whitespace between the comments to do that. Could do a skipWhitespace(file.text, comment.end) I suppose. But it is starting to get annoying that we have to do things like that. So it does feel to me that there should be an API that returns all trivia on top of which all the various helpers could be built.

Reply:
Thinking about it some more, skipWhitespace(file.text, comment.end) is fine and the code would come out similar to how we collapse consecutive decorators with a skipTrivia(file.text, decorator.end) to find the end of the collapsed decorators. We can probably get this feature cheaply with that and kick the can down the road for fancier trivia helpers or parsing :)

Reply:
Also totally fine to do this as a separate change / issue.

Reply:
Looks like we don't have skipWhitespace yet, but it would be trivial to add.

Reply:
Filed https://github.com/microsoft/cadl/issues/701